cli/src/main/kotlin/org/partiql/format/DotFormatter.kt

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at:
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package org.partiql.format
+
+import org.partiql.format.dot.DotGraph
+import org.partiql.format.dot.DotNodeId
+import org.partiql.format.dot.DotNodeShape
+import org.partiql.format.dot.DotNodeStmt
+import org.partiql.format.dot.digraph
+import org.partiql.lang.domains.PartiqlAst
+import org.partiql.lang.domains.PartiqlLogical
+import org.partiql.lang.domains.PartiqlLogicalResolved
+import org.partiql.lang.domains.PartiqlPhysical
+import org.partiql.pig.runtime.DomainNode
+import kotlin.reflect.KClass
+import kotlin.reflect.jvm.isAccessible
+
+internal object DotFormatter : NodeFormatter {
+
+    private val NODE_MAP = mutableMapOf<KClass<*>, Int>()
+    private var CLUSTER_COUNT = 0
+    private const val START_NODE_ID = "START"
+    private val START_NODE: DotNodeStmt = DotNodeStmt(DotNodeId(START_NODE_ID)).also { node ->
+        node + {
+            shape = DotNodeShape.DIAMOND
+            style = "bold,filled"
+            fillcolor = "yellow"
+            fontcolor = "black"
+        }
+    }
+
+    override fun format(input: DomainNode): String {
```

Comment:
Overall this looks good. The only comment is that instance members  `NODE_MAP` and  `CLUSTER_COUNT` should be reset at the beginning of `format(input: DomainNode)` since DotFormatter is a singleton. Otherwise there could be bugs upon re-use.

Reply:
Totally right. Thanks.